pkg/storage/utils/decomposedfs/tree/tree.go

```diff
@@ -757,50 +770,62 @@ func (t *Tree) Propagate(ctx context.Context, n *node.Node) (err error) {
 				}
 			}
 
-			if err := n.UnsetTempEtag(); err != nil {
+			if err := n.UnsetTempEtag(); err != nil && !xattrs.IsAttrUnset(err) {
 				sublog.Error().Err(err).Msg("could not remove temporary etag attribute")
 			}
 		}
 
 		// size accounting
-		if t.treeSizeAccounting {
-			// update the treesize if it differs from the current size
-			updateTreeSize := false
-
-			var treeSize, calculatedTreeSize uint64
-			calculatedTreeSize, err = calculateTreeSize(ctx, n.InternalPath())
+		if t.treeSizeAccounting && sizeDiff != 0 {
+			// lock node before reading treesize
+			nodeLock, err := filelocks.AcquireWriteLock(n.InternalPath())
 			if err != nil {
-				continue
+				return err
+			}
+			// always unlock node
+			releaseLock := func() {
+				// ReleaseLock returns nil if already unlocked
+				if err := filelocks.ReleaseLock(nodeLock); err != nil {
+					sublog.Err(err).Msg("Decomposedfs: could not unlock parent node")
+				}
 			}
+			defer releaseLock()
+
+			var newSize uint64
 
-			treeSize, err = n.GetTreeSize()
+			// read treesize
+			treeSize, err := n.GetTreeSize()
 			switch {
+			case xattrs.IsAttrUnset(err):
+				// fallback to calculating the treesize
+				newSize, err = calculateTreeSize(ctx, n.InternalPath())
+				if err != nil {
+					return err
+				}
 			case err != nil:
-				// missing attribute, or invalid format, overwrite
-				sublog.Debug().Err(err).Msg("could not read treesize attribute, overwriting")
-				updateTreeSize = true
-			case treeSize != calculatedTreeSize:
-				sublog.Debug().
-					Uint64("treesize", treeSize).
-					Uint64("calculatedTreeSize", calculatedTreeSize).
-					Msg("parent treesize is different then calculated treesize, updating")
-				updateTreeSize = true
+				return err
 			default:
-				sublog.Debug().
-					Uint64("treesize", treeSize).
-					Uint64("calculatedTreeSize", calculatedTreeSize).
-					Msg("parent size matches calculated size, not updating")
-			}
-
-			if updateTreeSize {
-				// update the tree time of the parent node
-				if err = n.SetTreeSize(calculatedTreeSize); err != nil {
-					sublog.Error().Err(err).Uint64("calculatedTreeSize", calculatedTreeSize).Msg("could not update treesize of parent node")
+				if sizeDiff > 0 {
```

Comment:
Alternative approach to get rid of the `if` (assuming that the `int64` cast is safe):

```
newSize = uint64(int64(newSize) + sizeDiff)
```

Reply:
but this allows handling bigger treesizes